core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala

```diff
@@ -160,7 +160,26 @@ abstract class BaseQuotaTest extends IntegrationTestHarness {
 }
 
 object QuotaTestClients {
-  def metricValue(metric: Metric): Double = metric.metricValue().asInstanceOf[Double]
+  // This method may be invoked when metrics are being updated. Reading metricValue
+  // during an update may cause ConcurrentModificationException or NullPointerException.
+  // Retry a few times with a very small delay. We want to allow reading throttle metrics
+  // without waiting for request to complete to avoid waiting for potentially large
+  // throttle times.
+  def metricValue(metric: Metric): Double = {
+    val maxRetries = 10
+    (1 to maxRetries).foreach { i =>
+      try {
+        return metric.metricValue().asInstanceOf[Double]
```

Comment:
Hmm, if this issue can happen in unit test, it will affect the JmxMetricReporter as well. So, perhaps we should fix this in Metric itself.

In 0.8.2, we have the following code in Sensor.java. I guess the idea is that if the reader/writer need synchronization on a stat, we will use this method to add a metric. That way, the reader/writer of the metric will be synchronized on the Sensor object. If this is the case, the right fix is probably to convert all usage of SampledStat to use an api like that to add a metric.

```
    public synchronized void add(CompoundStat stat, MetricConfig config) {
        this.stats.add(Utils.notNull(stat));
        for (NamedMeasurable m : stat.stats()) {
            KafkaMetric metric = new KafkaMetric(this, m.name(), m.stat(), config == null ? this.config : config, time);
            this.registry.registerMetric(metric);
            this.metrics.add(metric);
        }
    }
```

Reply:
@junrao Thanks for the review. I was wondering about JmxMetricReporter when I fixed the test. When we started reading metrics during the requests in the quota tests, tests started failing very frequently, but we haven't had any reports of exceptions in the metrics reporters. So I thought perhaps the issue was only with the tests. But you are right, we should fix the implementation, it may be just that we aren't testing this elsewhere with multiple threads. Your suggestion makes sense. I have updated the code and added a test.